### issue

airflow task를 브랜드 전체 개수를 코어 수로 나눠서 multiprocessing되도록 했는데, `AssertionError: daemonic processes are not allowed to have children` 에러가 발생한다.

https://leo-bb.tistory.com/87 여기에 정말 잘 정리되어있다.

데몬프로세스(=airflow task process)는 자식프로세스(=내가 정의한 프로세스)를 가질 수 없어서 발생하는 오류다. 에어플로우 컨트리뷰터가 "The best way for you to proceed will be to turn yout multiprocessing jobs into airflow tasks" 라고 했다고 한다.

그럼 두 가지 선택지가 있다.
1. 얌전히 컨트리뷰터의 말을 듣고`test1` 방법으로 리팩토링한다.(task 여러개 생성),
2. 지금의 multiprocessing 그대로 가고 민준님이 제안하셨던 VirtualenvOperator로 dag를 수정해서 가상환경에서 코드 동작되도록 해서 파이썬 인터프리터가 데몬프로세스 말고 그냥 새 프로세스를 생성하도록 한다.

2번부터 테스트해보자

--- a/brickstudy_ingestion/dags/viral/scrapper_brand.py
+++ b/brickstudy_ingestion/dags/viral/scrapper_brand.py
@@ -20,17 +20,17 @@ default_args = {
 
 def entrypoint():
     import logging
-    import multiprocessing
+    import multiprocess
     import uuid
     from src.scrapper.oliveyoung import Brand
-    from src.scrapper.utils import dict_partitioner
+    from src.scrapper.utils import dict_partitioner, write_local_as_json
 
-    CONCURRENCY_LEVEL = multiprocessing.cpu_count()
+    CONCURRENCY_LEVEL = multiprocess.cpu_count()
 
     def get_item(data: dict):
         brand = Brand(data)
         brand.crawl_items()
-        file_name = 'brand_item_data_' + uuid.uuid4()
+        file_name = 'brand_item_data_' + str(uuid.uuid4())
         write_local_as_json(data=brand.brand_metadata, file_name=file_name)
 
     try:
@@ -40,9 +40,8 @@ def entrypoint():
             data=brand.brand_metadata,
             level=CONCURRENCY_LEVEL
         )
-        with multiprocessing.Pool(CONCURRENCY_LEVEL) as p:
-            p.map(get_item, partitioned_data)
-        # get_item(partitioned_data)
+        with multiprocess.Pool(CONCURRENCY_LEVEL) as p:
+            p.map(get_item, list(partitioned_data.items()))
     except Exception as e:
         logging.error("***entrypoint error***", e)
         raise
@@ -58,7 +57,7 @@ with DAG(
         task_id='get_brand_metadata',
         python_version='3.7',
         system_site_packages=False,
-        requirements=["beautifulsoup4==4.9.3", "python-dotenv==0.19.0"],
+        requirements=["beautifulsoup4==4.9.3", "python-dotenv==0.19.0", "multiprocess"],
         python_callable=entrypoint
     )
 

--- a/brickstudy_ingestion/dags/viral/scrapper_brand.py
+++ b/brickstudy_ingestion/dags/viral/scrapper_brand.py
@@ -1,20 +1,16 @@
 from datetime import datetime
-import multiprocessing
-import uuid
-import logging
 
 from airflow import DAG
 from airflow.operators.python import PythonVirtualenvOperator
 
 from src.scrapper.oliveyoung import Brand
 from src.scrapper.utils import (
-    dict_partitioner,
+    # dict_partitioner,
     write_local_as_json,
 )
 
 
 DAG_ID = "bronze_viral_oliveyoung"
-CONCURRENCY_LEVEL = multiprocessing.cpu_count()
 
 default_args = {
     'owner': 'airflow',
@@ -23,6 +19,20 @@ default_args = {
 
 
 def entrypoint():
+    import logging
+    import multiprocessing
+    import uuid
+    from src.scrapper.oliveyoung import Brand
+    from src.scrapper.utils import dict_partitioner
+
+    CONCURRENCY_LEVEL = multiprocessing.cpu_count()
+
+    def get_item(data: dict):
+        brand = Brand(data)
+        brand.crawl_items()
+        file_name = 'brand_item_data_' + uuid.uuid4()
+        write_local_as_json(data=brand.brand_metadata, file_name=file_name)
+
     try:
         brand = Brand()
         brand.crawl_brand_metadata()
@@ -30,19 +40,12 @@ def entrypoint():
             data=brand.brand_metadata,
             level=CONCURRENCY_LEVEL
         )
-        # with multiprocessing.Pool(CONCURRENCY_LEVEL) as p:
-        #     p.map(get_item, partitioned_data)
-        get_item(partitioned_data)
+        with multiprocessing.Pool(CONCURRENCY_LEVEL) as p:
+            p.map(get_item, partitioned_data)
+        # get_item(partitioned_data)
     except Exception as e:
         logging.error("***entrypoint error***", e)
         raise
-
-
-def get_item(data: dict):
-    brand = Brand(data)
-    brand.crawl_items()
-    file_name = 'brand_item_data_' + uuid.uuid4()
-    write_local_as_json(data=brand.brand_metadata, file_name=file_name)
 
 
 with DAG(
@@ -53,9 +56,9 @@ with DAG(
 
     task1 = PythonVirtualenvOperator(
         task_id='get_brand_metadata',
-        python_version='3.8',
+        python_version='3.7',
         system_site_packages=False,
-        requirements=["beautifulsoup4==4.9.3"],
+        requirements=["beautifulsoup4==4.9.3", "python-dotenv==0.19.0"],
         python_callable=entrypoint
     )
 

--- a/brickstudy_ingestion/dags/viral/scrapper_brand.py
+++ b/brickstudy_ingestion/dags/viral/scrapper_brand.py
@@ -1,9 +1,10 @@
 from datetime import datetime
 import multiprocessing
 import uuid
+import logging
 
 from airflow import DAG
-from airflow.operators.python import PythonOperator
+from airflow.operators.python import PythonVirtualenvOperator
 
 from src.scrapper.oliveyoung import Brand
 from src.scrapper.utils import (
@@ -22,14 +23,19 @@ default_args = {
 
 
 def entrypoint():
-    brand = Brand()
-    brand.crawl_brand_metadata()
-    for partitioned_data in dict_partitioner(
-        data=brand.brand_metadata,
-        level=CONCURRENCY_LEVEL
-    ):
-        with multiprocessing.Pool(CONCURRENCY_LEVEL) as p:
-            p.map(get_item, partitioned_data)
+    try:
+        brand = Brand()
+        brand.crawl_brand_metadata()
+        partitioned_data = dict_partitioner(
+            data=brand.brand_metadata,
+            level=CONCURRENCY_LEVEL
+        )
+        # with multiprocessing.Pool(CONCURRENCY_LEVEL) as p:
+        #     p.map(get_item, partitioned_data)
+        get_item(partitioned_data)
+    except Exception as e:
+        logging.error("***entrypoint error***", e)
+        raise
 
 
 def get_item(data: dict):
@@ -45,8 +51,11 @@ with DAG(
     schedule_interval='@daily'
 ):
 
-    task1 = PythonOperator(
+    task1 = PythonVirtualenvOperator(
         task_id='get_brand_metadata',
+        python_version='3.8',
+        system_site_packages=False,
+        requirements=["beautifulsoup4==4.9.3"],
         python_callable=entrypoint
     )
 

--- a/docker-compose.yaml
+++ b/docker-compose.yaml
@@ -71,9 +71,9 @@ x-airflow-common:
     AIRFLOW__SCHEDULER__ENABLE_HEALTH_CHECK: 'true'
     # WARNING: Use _PIP_ADDITIONAL_REQUIREMENTS option ONLY for a quick checks
     # for other purpose (development, test and especially production usage) build/extend Airflow image.
-    _PIP_ADDITIONAL_REQUIREMENTS: ${_PIP_ADDITIONAL_REQUIREMENTS:- apache-airflow-providers-postgres}
+    _PIP_ADDITIONAL_REQUIREMENTS: ${_PIP_ADDITIONAL_REQUIREMENTS:-apache-airflow[postgres,virtualenv]}
     AIRFLOW__CORE__DAGS_FOLDER: /opt/airflow/brickstudy_ingestion/dags
-    PYTHONPATH: /opt/airflow/brickstudy_ingestion
+    PYTHONPATH: $PYTHONPATH:/opt/airflow/brickstudy_ingestion
   volumes:
     - ${AIRFLOW_PROJ_DIR:-.}/brickstudy_ingestion:/opt/airflow/brickstudy_ingestion
     - ${AIRFLOW_PROJ_DIR:-.}/logs:/opt/airflow/logs

